added images endpoint

Creepsy opened the pull request:
- added endpoint "recipes/:recipeId/images" for retrieving all images associated with the specified recipe
- added endpoint "/images/:uuid" for retrieving the specified image

Creepsy committed:
added images endpoint
- added endpoint "recipes/:recipeId/images" for retrieving all images associated with the specified recipe
- added endpoint "/images/:uuid" for retrieving the specified image

Creepsy committed:
refactoring & removed unused imports

Creepsy committed:
removed name shadowing & fixed formatter producing erroneous code

rehwinkel commented on src/Main.hs:
```diff
@@ -32,11 +34,11 @@ main :: IO ()
   createTables db
   adminSuccess <- createUser db "admin" (adminPassword config)
   when adminSuccess $ putStrLn "Created user \"admin\""
+  -- _ <- liftIO $ storeImageInDB db "some_image" (ImageRGB8 $ generateImage (curry $ const (PixelRGB8 10 10 10)) 100 100) 1
```
Bitte entfernen

Creepsy replied:
Resolved by 6bdaf97f2199963cc14258da8ffc9c0a22927913

rehwinkel commented on src/Images.hs:
```diff
@@ -0,0 +1,98 @@
+  toRow (ImageInfo imageUUID imageTitle fPath rId) = [toField . show $ imageUUID, toField imageTitle, toField fPath, toField rId]
+
+imageFolder :: FilePath
+imageFolder = "images/"
```
Bitte aus config auslesen

Creepsy replied:
Resolved by e75f8bc5613b05785fa0ad2992cda0b4fa7b4aab

rehwinkel commented on src/Images.hs:
```diff
@@ -0,0 +1,98 @@
+    "CREATE TABLE IF NOT EXISTS images (\
+    \ uuid VARCHAR(36) PRIMARY KEY NOT NULL,\
+    \ title VARCHAR(40) NOT NULL,\
+    \ filePath VARCHAR(100) NOT NULL,\
```
Bitte entfernen, da UUID den pfad angibt

Creepsy replied:
Resolved by 6bdaf97f2199963cc14258da8ffc9c0a22927913

rehwinkel commented on src/Images.hs:
```diff
@@ -0,0 +1,98 @@
+data ImageInfo = ImageInfo
+  { uuid :: UUID,
+    title :: String,
+    filePath :: FilePath,
```
Auch hier filePath entfernen

Creepsy replied:
Resolved by 6bdaf97f2199963cc14258da8ffc9c0a22927913

rehwinkel commented on src/Images.hs:
```diff
@@ -0,0 +1,98 @@
+imageFolder :: FilePath
+imageFolder = "images/"
+
+getImageFromDB :: Connection -> UUID -> IO (Occurence DynamicImage)
```
Da UUID PrimaryKey ist sollte hier eher Maybe als Occurence sein.

Creepsy replied:
Resolved by 6bdaf97f2199963cc14258da8ffc9c0a22927913

rehwinkel commented on src/Images.hs:
```diff
@@ -0,0 +1,98 @@
+storeImageInDB :: Connection -> String -> DynamicImage -> RecipeID -> IO UUID
+storeImageInDB db imageTitle image rId = do
+  randomUUID <- nextRandom
+  let fPath = imageFolder ++ show randomUUID ++ ".jpg"
```
hier fehlt u.U. ein "/", wenn der Pfad konfigurierbar ist. Gibt es kein Path-Concat in Haskell?

Creepsy replied:
Resolved by e75f8bc5613b05785fa0ad2992cda0b4fa7b4aab

rehwinkel commented on src/Images.hs:
```diff
@@ -0,0 +1,98 @@
+    Many -> error "Multiple images with the same UUID found"
+
+handleGetAssociatedImageUUIDs :: Connection -> ActionM ()
+handleGetAssociatedImageUUIDs db = do
```
Das hier ist für's erste OK, sollte aber bei gelegenheit in getRecipe integriert werden

Creepsy replied:
No changes made

Creepsy replied:
marked as resolved

Creepsy committed:
applied some requested changes of merge request
- Removed comment in main file
- removed file path from database & code
- replaced occurence with maybe

Creepsy committed:
image folder is now being read from config

Creepsy committed:
refactoring & formatting

rehwinkel commented on src/Images.hs:
```diff
@@ -0,0 +1,92 @@
+storeImageOnDisk fPath = BS.writeFile fPath . imageToJpg 50
+
+imagePath :: FilePath -> UUID -> FilePath
+imagePath imageFolder imageUUID = imageFolder </> show imageUUID <.> ".jpg"
```
sicher dass das richtig ist? oder muss das "." bei ".jpg" weg?

Creepsy replied:
Doesn't matter according to the documentation. Changed by d1ab51931033a71d4ad34948b1ba13ad35e84793 for better readability

rehwinkel commented on src/Images.hs:
```diff
@@ -0,0 +1,92 @@
+getImageFromDisk imageFolder imageUUID = either (const Nothing) Just <$> (readImage . imagePath imageFolder $ imageUUID)
+
+storeImageOnDisk :: FilePath -> DynamicImage -> IO ()
+storeImageOnDisk fPath = BS.writeFile fPath . imageToJpg 50
```
Wenn 50 die Quality ist, muss das mind. 80 sein. 50 ist viel zu wenig. 80-90

Creepsy replied:
Quality changed to 80 by d1ab51931033a71d4ad34948b1ba13ad35e84793 (Maybe add this to the config in the future?)

rehwinkel commented on src/Images.hs:
```diff
@@ -0,0 +1,92 @@
+    Image:
+      uuid:     UUID of the image
+      title:    Title of the image
+      filePath: Filepath to the image
```
kommentar wurde nicht angepasst

Creepsy replied:
Comment removed by d1ab51931033a71d4ad34948b1ba13ad35e84793

Creepsy committed:
applied change requests
